### ISSUE TITLE
Undo

### ISSUE THREAD

**Issue by liaprins-czi** [napari/product-heuristics-2020#26]
Lack of undo is very prominently felt (especially in painting labels).
__________________________________________________________
**Recommendation**

Determine best interactions for user to enact undo/redo (e.g. use universal keybindings + “File > Undo [insert action to be undone]” from within menubar, but also consider onscreen buttons). Implement.

Moderate issue
●●○   Medium design work
▼▼▼   High development work

**Comment by goanpeca** [comment]
Could we have a more detailed list of the different actions where we would like to implement undo/redo, and maybe define some priorities @sofroniewn ?

**Comment by sofroniewn** [comment]
Yup, will get on this - at a high level, I think there will be things which change the actual data (like painting labels/ adding points, or one day an analysis routine from a plugin) and things which change how the data gets displayed (colormaps, rendering modes, moving the sliders).

I think ultimately we'd like all actions to be undoable and to have a nice history feature like photoshop has, but I can work on prioritizing use cases.

I'd like to us to keep this in mind as we work on functionality like our action manager and maybe tools like macro-generation as I see ways in which they are connected.

We created this issue on the main repo a while back that had some ideas at the time https://github.com/napari/napari/issues/474, but we never begun any implementation. Just putting it here as a reference, our thinking may have evolved since then.

More coming!

**Comment by liaprins-czi** [comment]
G-doc with research:
https://docs.google.com/document/d/1TDlZV874qvt1pFFZbCFmaYJ1-w1_13EoejdD7uEoe8U/edit#

Old undo issue in main napari repo #474:
https://github.com/napari/napari/issues/474

New undo/redo issue in main napari repo #2579:
https://github.com/napari/napari/issues/2579

Design steps:
- [x] Read existing GH issues to understand context better
- [x] Research + document design best practices, examples for this issue (including how other imaging platforms handle undo-redo)
- [ ] Define scope of design for first iteration / implementation
- [ ] Sketches to iterate solutions
- [ ] Narrow down solutions
- [ ] Recreate narrowed-down solutions in Figma
- [ ] Narrow down to final design
- [ ] Document final design

**Comment by liaprins-czi** [comment]
@goanpeca @ppwadhwa Temporarily un-assigning you from this while I work on design for it; I will re-assign to you all once design is ready (and not that you need to start working on it as soon as design is ready, that should not change the priority per se)

**Comment by psobolewskiPhD** [comment]
Undo does work in Labels, but isn't very discoverable due to lack of Edit menu.
Doesn't work in Shapes...
So this definitely is still a high priority to improve!
